src/Aspire.Hosting/DistributedApplicationBuilder.cs

```diff
@@ -226,20 +233,62 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
         }
 
         // Publishing support
-        _innerBuilder.Services.AddLifecycleHook<Http2TransportMutationHook>();
-        _innerBuilder.Services.AddLifecycleHook<DashboardManifestExclusionHook>();
+        Eventing.Subscribe<BeforeStartEvent>(MutateHttp2TransportAsync);
         _innerBuilder.Services.AddKeyedSingleton<IDistributedApplicationPublisher, ManifestPublisher>("manifest");
 
+        Eventing.Subscribe<BeforeStartEvent>(ExcludeDashboardFromManifestAsync);
+
         // Overwrite registry if override specified in options
         if (!string.IsNullOrEmpty(options.ContainerRegistryOverride))
         {
-            _innerBuilder.Services.AddLifecycleHook<ContainerRegistryHook>();
+            Eventing.Subscribe<BeforeStartEvent>((e, ct) => UpdateContainerRegistryAsync(options));
         }
 
         _innerBuilder.Services.AddSingleton(ExecutionContext);
         LogBuilderConstructed(this);
     }
 
+    private Task MutateHttp2TransportAsync(BeforeStartEvent beforeStartEvent, CancellationToken cancellationToken)
+    {
+        foreach (var resource in Resources)
+        {
+            var isHttp2Service = resource.Annotations.OfType<Http2ServiceAnnotation>().Any();
+            var httpEndpoints = resource.Annotations.OfType<EndpointAnnotation>().Where(sb => sb.UriScheme == "http" || sb.UriScheme == "https");
+            foreach (var httpEndpoint in httpEndpoints)
+            {
+                httpEndpoint.Transport = isHttp2Service ? "http2" : httpEndpoint.Transport;
+            }
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private Task UpdateContainerRegistryAsync(DistributedApplicationOptions options)
+    {
+        var resourcesWithContainerImages = Resources.SelectMany(
+            r => r.Annotations.OfType<ContainerImageAnnotation>()
+                              .Select(cia => new { Resource = r, Annotation = cia })
+            );
+
+        foreach (var resourceWithContainerImage in resourcesWithContainerImages)
+        {
+            resourceWithContainerImage.Annotation.Registry = options.ContainerRegistryOverride;
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private Task ExcludeDashboardFromManifestAsync(BeforeStartEvent beforeStartEvent, CancellationToken cancellationToken)
+    {
+        // When developing locally, exclude the dashboard from the manifest. This only affects our playground projects in practice.
+        if (Resources.SingleOrDefault(r => StringComparers.ResourceName.Equals(r.Name, KnownResourceNames.AspireDashboard)) is { } dashboardResource)
```

Comment:
Seeing this it makes we wonder if we should pass the model as part of the event data for parity.

Reply:
updated events.